src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1893,7 +1893,21 @@ Value OpFusionHelper::emitFuseOps(
     defOpResult =
         emitScalar(rewriter, loc, useOp, currentElementType, inputValues);
     defOp = useOp;
-    alloc = defOp->getResult(0);
+
+    // Explanation:
+    // The variable, alloc, is used to initialize ONNXBroadcastOpShapeHelper.
+    // The output of defOp should have the same shape as the alloc, which is
+    // one of the prerequisits for fusion.
+    // However, they may have different element type for some ops, such as
+    // comparison and cast.
+    // The issue is that ONNXBroadcastOpShapeHelper.computeShape() may use
+    // this parameter to generate some code (memref.dim) that is not really
+    // needed here. Due to this live user, the original op can not be erased.
+    // This error occurs when there are more than one op with dynamic dim
+    // to be fused.
+    // Luckily, the ShapeHelper does not check the element type.
+    // Therefore, alloc is used for all the fused op.
+    // alloc = defOp->getResult(0);
```

Comment:
@chentong319 is it a temporary fix or do you think the right approach would be checking multiple ops with dynamic dim as you said `This error occurs when there are more than one op with dynamic dim to be fused`?

Also, since `alloc = defOp->getResult(0);` was removed, it's weird to have comments at the end of the function. Perhaps move the comment somewhere up.

Reply:
I think the fix is the right approach. In ONNXBroadcastOpShapeHelper, only the shape of tensor is consider, not the element type of the tensor. Therefore, alloc can be used for different op, as long as it guaranteed that it has the same shape for that operand, which has only scalar value not tensor value with fusion. That's the tricky part of the code gen.
I will move  the comment to the operands construction and add `const`.